addon/commands/insert-newLi-command.ts

```diff
@@ -17,261 +17,334 @@ export default class InsertNewLiCommand extends Command {
-        deleteNode = false;
-      } else {
-        // we are the root
-        return;
       }
     }
+    toBeRemoved.forEach(node => { node.remove(); });
   }
 
-  private deleteRight(node: ModelNode): void {
-    //make sure the first node is deleted
-    let deleteNode = true;
-    while (node) {
-      //stop if we reached the parent li
-      if (ModelNode.isModelElement(node) && node.type === "li") {
-        return;
+  //this function finds the first node to delete if position.nodeAfter/position.nodeBefore don't return anything usefull (ie: at the end of a child element)
+  //this will look for the firt sibling of a parent
+  findClosestNode(position: ModelPosition, direction: string) {
+    let searchVar;
+    if (direction == "right") {
+      searchVar = position.nodeBefore();
+    }
+    else if (direction == "left") {
+      searchVar = position.nodeAfter();
+    }
+
+    if (searchVar && searchVar.parent) {
+      searchVar = searchVar.parent;
+    }
+
+    while (searchVar) {
+      if (ModelNode.isModelElement(searchVar) && searchVar.type === "li") {
+        break;
       }
-      if (deleteNode) {
-        this.model.removeModelNode(node);
+      else if (direction == "right" && searchVar.nextSibling) {
+        searchVar = searchVar.nextSibling;
+        return searchVar;
       }
-      //remove all siblings to the left but keep direct parents
-      if (node.nextSibling) {
-        node = node.nextSibling;
-        deleteNode = true;
-      } else if (node.parent) {
-        node = node.parent;
-        deleteNode = false;
-      } else {
-        return;
+      else if (direction == "left" && searchVar.previousSibling) {
+        searchVar = searchVar.previousSibling;
+        return searchVar;
+      }
+      else if (searchVar.parent) {
+        searchVar = searchVar.parent;
       }
     }
+    return null;
   }
 
-  private deleteSelection(first: ModelNode, last: ModelNode) {
-    //make sure the first node is deleted
-    let node = last;
-    let deleteNode = true;
-    while (node) {
-      //stop if we reached the first node
-      if (node == first) {
-        break;
+  // core algorithm
+  // this basically splits a li
+  // like so:
+  //
+  // <li>
+  //  <div class="some stuff">
+  //   some text
+  //       ^cursor
+  //  </div>
+  // </li>
+  //
+  // turns into:
+  //
+  // <li>
+  //  <div class="some stuff">
+  //   some
+  //  </div>
+  // </li>
+  // <li>
+  //  <div class="some stuff">
+  //   text
+  //  ^cursor
+  //  </div>
+  // </li>
+  //
+  // when a non-collapsed selection is preset
+  // it basically deletes all the highlighted stuff
+  // and turns it into a collapsed case
+  // if stuff is unclear please ask sergey.andreev@redpencil.io
+  insertNewLi(position: ModelPosition): void {
+
+    //split the nodes at position
+    position.split();
+
+    //variable initialization
+    let parentLi;
+    let parentUl;
+    let parentLiPos;
+    let toBeInserted: ModelNode[] = [];
```

Comment:
toBeInserted and newLiNodes can be const and defined closer to where they are used since they are only used in the upcoming else branch